kit/popups/src/Menu/UIMenuContainer.tsx

```diff
@@ -0,0 +1,189 @@
+import { UIConstant } from '../constants';
+import { ShadowView } from '../ShadowView';
+
+type Location = {
+    left: number;
+    top: number | undefined;
+    bottom: number | undefined;
+};
+type Size = {
+    width: number;
+    height: number;
+};
+type Dimensions = {
+    x: number;
+    y: number;
+    width: number;
+    height: number;
+};
+
+const initialSize: Size = {
+    width: 0,
+    height: 0,
+};
+
+function useTargetDimensions(
+    visible: boolean,
+    targetRef: React.RefObject<NativeMethods>,
+    windowDimensions: ScaledSize,
+): Dimensions | null {
+    const [dimensions, setDimensions] = React.useState<Dimensions | null>(null);
+
+    React.useEffect(() => {
+        if (visible) {
+            targetRef.current?.measure(
+                (
+                    _x: number,
+                    _y: number,
+                    width: number,
+                    height: number,
+                    pageX: number,
+                    pageY: number,
+                ) => {
+                    setDimensions({
+                        x: pageX,
+                        y: pageY,
+                        width,
+                        height,
+                    });
+                },
+            );
+        } else {
+            setDimensions(null);
+        }
+    }, [targetRef, visible, windowDimensions]);
+
+    return dimensions;
+}
+
+function getBoundaries(windowDimensions: ScaledSize, menuSize: Size) {
+    return {
+        top: UILayoutConstant.contentOffset,
+        right: windowDimensions.width - UILayoutConstant.contentOffset - menuSize.width,
+        bottom: windowDimensions.height - UILayoutConstant.contentOffset - menuSize.height,
+        left: UILayoutConstant.contentOffset,
+    };
+}
+
+function useMenuLocation(
+    targetDimensions: Dimensions | null,
+    windowDimensions: ScaledSize,
+    menuSize: Size,
+): Location | null {
+    return React.useMemo(() => {
+        if (!menuSize.height || !menuSize.width || !targetDimensions) {
+            return null;
+        }
+
+        let top: number | undefined = targetDimensions.y + targetDimensions.height;
+        let bottom: number | undefined;
+        let left: number = targetDimensions.x;
+
+        const boundaries = getBoundaries(windowDimensions, menuSize);
+
+        if (left > boundaries.right) {
+            left = boundaries.right;
+        }
+        if (left < boundaries.left) {
+            left = boundaries.left;
+        }
+
+        if (top > boundaries.bottom) {
+            top = undefined;
+            bottom = windowDimensions.height - targetDimensions.y;
+        }
+
+        return {
+            left,
+            top,
+            bottom,
+        };
+    }, [targetDimensions, windowDimensions, menuSize]);
+}
+
+export function UIMenuContainer({ children, visible, targetRef, onClose }: UIMenuContainerProps) {
+    const theme = useTheme();
+    const windowDimensions = useWindowDimensions();
```

Comment:
Frankly I don't like that all those computations are done outside of a `portal`, that means that if we had 100 menus, we would have 100 listeners hanging in memory.

What do you think to wrap it all with a `UIMenuContainerContent` component?

Reply:
Great idea!